### DISCUSSION TITLE
Possible bug in computation of second time derivative

### DISCUSSION THREAD

**Discussion by NateAM** [idaholab/moose#26128] [Q&A General]
I think I may have found a bug in the computation of the second time derivative of an array kernel's variable. If I define a simple kernel of the form:

ArrayTestKernel.h

    #pragma once

    #include "ArrayKernel.h"

    class ArrayTestKernel : public ArrayKernel{

        public:
            ArrayTestKernel( const InputParameters & params );

            static InputParameters validParams( );

        protected:

            const ArrayVariableValue & _u_dot;

            const ArrayVariableValue & _u_dotdot;

            virtual void computeQpResidual( RealEigenVector &residual ) override;

    };

ArrayTestKernel.C

    #include "ArrayTestKernel.h"

    registerMooseObject( "fireflyApp", ArrayTestKernel );

    InputParameters ArrayTestKernel::validParams( ){

        InputParameters params = ArrayKernel::validParams( );

        return params;

    }

    ArrayTestKernel::ArrayTestKernel( const InputParameters & parameters )
        : ArrayKernel( parameters ),
        _u_dot( _var.uDot( ) ),
        _u_dotdot( _var.uDotDot( ) ){

    }

    void ArrayTestKernel::computeQpResidual( RealEigenVector & residual ){

        std::cerr << "_u_dot:\n" << _u_dot[ _qp ] << "\n";
        std::cerr << "_u_dotdot:\n" << _u_dotdot[ _qp ] << "\n";

        return;

    }

If I run this with Newmark-Beta integration to enable the second time derivative using

my_test.i

    [Mesh]
      type = GeneratedMesh
      dim = 1
      nx = 3
    []

    [Variables]
      [u]
        components = 2
      []
    []

    [Kernels]
      [test]
        type = ArrayTestKernel
        variable = u
        use_displaced_mesh = false
      []
      [u_diffusion]
        type = ArrayDiffusion
        variable = u
        diffusion_coefficient = u_dc
        use_displaced_mesh = false
      []
      [u_time]
        type = ArrayTimeDerivative
        variable = u
        use_displaced_mesh = false
      []
    []

    [ICs]
      [u]
        type = ArrayFunctionIC
        variable = u
        function = '.3*(-x+y-0.1*z) 0.19*(-x+y+z)'
      []
    []

    [Materials]
      [u_dc]
        type = GenericConstantArray
        prop_name = u_dc
        prop_value = '1 1'
      []
    []

    [Preconditioning]
      [./smp]
        type = SMP
        full = true
      [../]
    []

    [Executioner]
      type = Transient
      dt = 0.1
      dtmin = 0.1
      num_steps = 3
      scheme = newmark-beta
      solve_type = 'PJFNK'
      petsc_options = '-snes_test_jacobian -snes_test_jacobian_view'
      petsc_options_iname = '-pc_type'
      petsc_options_value = 'lu'
    []

    [Outputs]
      exodus = true
    []

Then I get that `_u_dot` is exactly the same as `_u_dotdot` for all increments and iterations. This same approach works okay for non-array variables. Can anyone see where I've gone awry here or is this actually a problem in MOOSE?

**Comment by GiudGiud** [comment]
Hello

I think you may be the first one to look at array variables and second time derivatives.
Can you please try a simple time integrator instead of NewmarkBeta first?

Guillaume

**Reply by NateAM** [reply]
@GiudGiud Do any of the other implicit time integrators make the second time derivatives available?

**Reply by GiudGiud** [reply]
Does not look like it. CentralDifference is explicit iirc and has 2nd time derivatives

**Reply by NateAM** [reply]
Okay, so I don't *think* I can use anything other than Newmark-Beta. I tried to use `setUDotDotRequested(true)` in the constructor of `ArrayKernelTest` but that fails.

The only classes that I could find that request the second time derivative are time integrators so I don't have much to follow.

**Reply by GiudGiud** [reply]
I ll try to drum up something with BDF2

**Reply by GiudGiud** [reply]
Seems to happen as well with a simple backwards difference for the second time derivative with BDF2
https://github.com/GiudGiud/moose/tree/PR_second

**Reply by GiudGiud** [reply]
The computation seems to be correct for BDF2. But the kernel prints the wrong value. We're missing something.
@hugary1995 @cbolisetti any idea

**Reply by cbolisetti** [reply]
We still calculate second time derivative using NewmarkBetaAccelAux auxkernel in a lot of cases, instead of using the time integrator. It should work for any variable, I think. Its probably the simplest way for your case.
